data/src/opaque.rs

```diff
@@ -7,10 +8,47 @@ use std::sync::Arc;
 use downcast_rs::{impl_downcast, Downcast};
 use dyn_hash::DynHash;
 
-pub trait OpaquePayload: DynHash + Send + Sync + Debug + Display + Downcast {}
+pub trait OpaquePayload: DynHash + Send + Sync + Debug + Display + Downcast {
+
+    fn metadata(&self) -> Option<Box<dyn OpaqueMetadata>> {
```

Comment:
do we even need this ? Or is it up to the code instantianting the TypedFact to manage the OpaquePayload directly ? Asking because it does not look like we're using it anywhere in the PR.

Also it makes it a bit murky in the case where we have several OpaquePayload in a tensor. Better leave it entirely to the TypeFact instantiator, i think.

Reply:
The only situation, I see this function call it is when we receive a tensor in the `eval` and we want to get the metadata. But we could consider that in such situation, the user needs to downcast the opaque tensor to a concrete type and build the metadata itself ? WDYT ?

Reply:
Well, i just don't think we should assume the metadata should be constructed from the Opaque. If we build a TypedFact with metadata, we will just explicitely add it. This solves the problem where we have a tensor with multiple opaque inside: they could potentially generate different metadata, but the fact only stores one.

So yeah, If the metadata is still helpful at the `eval` stage, its code will just generate it again from the concrete Opaque type.